pkg/storage/azure/azure.go

```diff
@@ -430,53 +418,42 @@ func (d *driver) CreateStorage(cr *imageregistryv1.Config) error {
 		}
 
 		if d.Config.AccountName == "" {
-			infra, err := util.GetInfrastructure(d.Listers)
-			if err != nil {
-				util.UpdateCondition(cr, imageregistryv1.StorageExists, operatorapiv1.ConditionFalse, storageExistsReasonConfigError, fmt.Sprintf("Unable to get infrastructure resource: %s", err))
-				return err
-			}
-
 			const maxAttempts = 10
 			var lastErr error
 			for i := 0; i < maxAttempts; i++ {
 				accountName := generateAccountName(infra.Status.InfrastructureName)
-				if err := d.createStorageAccount(storageAccountsClient, cfg.ResourceGroup, accountName, cfg.Region); err != nil {
-					if e, ok := err.(azblob.StorageError); ok {
-						switch e.ServiceCode() {
-						case azblob.ServiceCodeContainerAlreadyExists:
-							klog.Warningf("unable to create storage container: %s", err)
-							lastErr = err
-						default:
-							return err
-						}
-					} else {
+				result, err := d.accountExists(storageAccountsClient, accountName)
+				if err != nil {
+					return err
+				}
+				if *result.NameAvailable {
+					if err := d.createStorageAccount(storageAccountsClient, cfg.ResourceGroup, accountName, cfg.Region); err != nil {
 						return err
 					}
+					d.Config.AccountName = accountName
+					cr.Status.StorageManaged = true
+					cr.Status.Storage.Azure = d.Config.DeepCopy()
+					cr.Spec.Storage.Azure = d.Config.DeepCopy()
+					break
 				}
-				d.Config.AccountName = accountName
-				cr.Status.StorageManaged = true
-				cr.Status.Storage.Azure = d.Config.DeepCopy()
-				cr.Spec.Storage.Azure = d.Config.DeepCopy()
-				break
 			}
 			if d.Config.AccountName == "" {
 				util.UpdateCondition(cr, imageregistryv1.StorageExists, operatorapiv1.ConditionFalse, storageExistsReasonAzureError, fmt.Sprintf("Unable to create storage account: %s", lastErr))
 				return fmt.Errorf("attmpts to create storage account failed, last error: %s", lastErr)
 			}
 		} else {
 			// TODO: do we need to create a storage account if we are provided with its name?
-			err = d.createStorageAccount(storageAccountsClient, cfg.ResourceGroup, d.Config.AccountName, cfg.Region)
+			result, err := d.accountExists(storageAccountsClient, d.Config.AccountName)
 			if err != nil {
-				if _, ok := err.(*errNameNotAvailable); !ok {
+				return err
+			}
+			if *result.NameAvailable {
+				if err = d.createStorageAccount(storageAccountsClient, cfg.ResourceGroup, d.Config.AccountName, cfg.Region); err != nil {
```

Comment:
We also need to mark that storage is managed in this situation. My understanding is that if we are creating the storage account, we are creating the underlying storage, too.

Reply:
updated in current push